src/frontend/src/handler/create_source.rs

```diff
@@ -302,12 +302,71 @@ fn get_name_strategy_or_default(name_strategy: Option<AstString>) -> Result<Opti
     }
 }
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum CreateSourceType {
+    SharedCdc,
+    /// e.g., shared Kafka source
+    SharedNonCdc,
+    NonShared,
+    /// create table with connector
+    Table,
+}
+
+impl CreateSourceType {
+    pub fn from_with_properties(
+        session: &SessionImpl,
+        with_properties: &impl WithPropertiesExt,
+    ) -> Self {
+        if with_properties.is_shareable_cdc_connector() {
+            CreateSourceType::SharedCdc
+        } else if with_properties.is_shareable_non_cdc_connector()
+            && session
+                .env()
+                .streaming_config()
+                .developer
+                .enable_shared_source
+            && session.config().streaming_use_shared_source()
+        {
+            CreateSourceType::SharedNonCdc
+        } else {
+            CreateSourceType::NonShared
+        }
+    }
+
+    pub fn is_shared(&self) -> bool {
+        matches!(
+            self,
+            CreateSourceType::SharedCdc | CreateSourceType::SharedNonCdc
+        )
+    }
+}
+
 /// Resolves the schema of the source from external schema file.
 /// See <https://www.risingwave.dev/docs/current/sql-create-source> for more information.
 ///
 /// Note: the returned schema strictly corresponds to the schema.
 /// Other special columns like additional columns (`INCLUDE`), and `row_id` column are not included.
-pub(crate) async fn bind_columns_from_source(
+pub async fn bind_columns_from_source(
+    session: &SessionImpl,
+    format_encode: &FormatEncodeOptions,
+    with_properties: Either<&WithOptions, &WithOptionsSecResolved>,
+    create_source_type: CreateSourceType,
+) -> Result<(Option<Vec<ColumnCatalog>>, StreamSourceInfo)> {
+    let (columns_from_resolve_source, mut source_info) =
+        if create_source_type == CreateSourceType::SharedCdc {
+            bind_columns_from_source_for_cdc(session, format_encode)?
+        } else {
+            bind_columns_from_source_for_non_cdc(session, format_encode, with_properties).await?
```

Comment:
- Renamed old `bind_columns_from_source` to `bind_columns_from_source_for_non_cdc`. Make it private. Previously it created a `mut StreamSourceInfo`
- Add new `bind_columns_from_source`, and put the code modifying `StreamSourceInfo` inside.